consensus/src/sync/live/block_queue/mod.rs

```diff
@@ -319,27 +316,91 @@ impl<N: Network> BlockQueue<N> {
+                vec![head_hash]
+            };
             // Prepend our current head hash.
-            let block_locators = vec![head_hash].into_iter().chain(block_locators).collect();
+            let block_locators = init_block_locators
+                .into_iter()
+                .chain(block_locators)
+                .collect();
 
             // FIXME Send missing blocks request to the peer that announced the block (first).
             self.request_component
-                .request_missing_blocks(block_hash, block_locators);
+                .request_missing_blocks(block_number, block_hash, block_locators);
         } else {
             log::error!(start_block = %head_hash, count = head_height - macro_height, "Couldn't get blocks")
         }
     }
 
     /// Handles missing blocks that were received.
-    fn handle_missing_blocks(&mut self, blocks: Vec<Block>) -> Option<QueuedBlock<N>> {
+    fn handle_missing_blocks(
+        &mut self,
+        target_block_number: u32,
+        target_hash: Blake2bHash,
+        blocks: Vec<Block>,
+    ) -> Option<QueuedBlock<N>> {
         if blocks.is_empty() {
             log::debug!("Received empty missing blocks response");
             return None;
         }
 
-        // FIXME Sanity-check blocks
+        // Checks that the first block was part of the chain or block locators.
+        let first_block = blocks.first()?;
+        if !self
+            .blockchain
+            .read()
+            .contains(first_block.parent_hash(), false)
+            && !self.blocks_pending_push.contains(first_block.parent_hash())
+            && !self
+                .buffer
+                .get(&first_block.block_number())
+                .map(|blocks| blocks.contains_key(first_block.parent_hash()))
+                .unwrap_or(false)
+        {
+            log::error!("Received invalid chain of missing blocks (first block not in chain or block locators).");
+            return None;
+        }
+        if first_block.block_number() > target_block_number {
+            log::error!("Received invalid chain of missing blocks (first block > target).");
+            return None;
+        }
+
+        // Check that the chain of missing blocks is valid.
+        let mut previous = first_block;
+        for block in blocks.iter().skip(1) {
+            if block.block_number() == previous.block_number() + 1
+                && block.block_number() <= target_block_number
+                && block.parent_hash() == &previous.hash()
+            {
+                previous = block;
+            } else {
+                log::error!("Received invalid chain of missing blocks.");
+                return None;
+            }
+        }
+
+        let last_block = blocks.last()?;
+        let block_hash = last_block.hash();
+        if last_block.block_number() == target_block_number && block_hash != target_hash {
+            log::error!("Received invalid missing blocks (invalid target block).");
```

Comment:
Maybe log `last_block.block_number()`, `last_block.block_number()`, `block_hash` and `target_hash`?